ksqldb-engine/src/main/java/io/confluent/ksql/utilization/PersistentQuerySaturationMetrics.java

```diff
@@ -0,0 +1,334 @@
+      return kafkaStreams.metrics().values().stream()
+          .filter(m -> m.metricName().group().equals(STREAMS_THREAD_METRICS_GROUP))
+          .collect(Collectors.groupingBy(
+              m -> m.metricName().tags().get(THREAD_ID),
+              Collectors.toMap(m -> m.metricName().name(), m -> m, (a, b) -> a)
+          ));
+    }
+
+    private Optional<Double> measure(
+        final Instant now,
+        final KafkaStreams kafkaStreams,
+        final MetricsReporter reporter
+    ) {
+      final Map<String, Map<String, Metric>> byThread = metricsByThread(kafkaStreams);
+      Optional<Double> saturation = Optional.of(0.0);
+      for (final Map.Entry<String, Map<String, Metric>> entry : byThread.entrySet()) {
+        final String threadName = entry.getKey();
+        final Map<String, Metric> metricsForThread = entry.getValue();
+        if (!metricsForThread.containsKey(STREAMS_TOTAL_BLOCKED_TIME)
+            || !metricsForThread.containsKey(STREAMS_THREAD_START_TIME)) {
+          LOGGER.info("Missing required metrics for thread: {}", threadName);
+          continue;
+        }
+        final double totalBlocked =
+            (double) metricsForThread.get(STREAMS_TOTAL_BLOCKED_TIME).metricValue();
+        final long startTime =
+            (long) metricsForThread.get(STREAMS_THREAD_START_TIME).metricValue();
+        final ThreadSaturation threadSaturation = perThreadSaturation.computeIfAbsent(
+            threadName,
+            k -> {
+              LOGGER.debug("Adding saturation for new thread: {}", k);
+              return new ThreadSaturation(threadName, startTime, window, sampleMargin);
+            }
+        );
+        LOGGER.debug("Record thread {} sample {} {}", threadName, totalBlocked, startTime);
+        final BlockedTimeSample blockedTimeSample = new BlockedTimeSample(now, totalBlocked);
+        final Optional<Double> measured = threadSaturation.measure(now, blockedTimeSample);
+        LOGGER.debug(
+            "Measured value for thread {}: {}",
+            threadName,
+            (measured.map(Object::toString).orElse(""))
+        );
+        measured.ifPresent(m -> reportThreadSaturation(now, m, threadName, reporter));
+        saturation = compareSaturation(saturation, measured) > 0 ? saturation : measured;
+      }
+      saturation.ifPresent(s -> reportQuerySaturation(now, s, queryId, reporter));
+      for (final String threadName
+          : Sets.difference(new HashSet<>(perThreadSaturation.keySet()), byThread.keySet())) {
+        perThreadSaturation.remove(threadName);
+        reporter.cleanup(QUERY_THREAD_SATURATION, ImmutableMap.of(THREAD_ID, threadName));
+      }
+      return saturation;
+    }
+
+    private void cleanup(final MetricsReporter reporter) {
+      for (final String threadName : perThreadSaturation.keySet()) {
+        reporter.cleanup(QUERY_THREAD_SATURATION, ImmutableMap.of(THREAD_ID, threadName));
```

Comment:
Are the extra thread metrics not already cleaned up at the end of `KafkaStreamsSaturation#measure`?

Reply:
we need to clean up the thread metrics in the case a thread restarts, or in the case an app is closed. This method is called in the latter case.